src/uct/ib/rdmacm/rdmacm_cm_ep.c

```diff
@@ -192,27 +314,69 @@ static ucs_status_t uct_rdmacm_cm_create_dummy_qp(struct rdma_cm_id *id,
     return UCS_OK;
 }
 
-static ucs_status_t
-uct_rdamcm_cm_ep_set_qp_num(struct rdma_conn_param *conn_param,
-                            uct_rdmacm_cm_ep_t *cep)
+static ucs_status_t uct_rdamcm_cm_ep_create_qpn(uct_rdmacm_cm_ep_t *cep,
+                                                uint32_t* qpn)
 {
     struct ibv_cq *cq;
     ucs_status_t status;
+    bool supported;
 
-    status = uct_rdmacm_cm_get_cq(uct_rdmacm_cm_ep_get_cm(cep), cep->id->verbs,
-                                  &cq);
+    status = uct_rdamcm_cm_ep_is_reserved_qpn_supported(cep, &supported);
     if (status != UCS_OK) {
         return status;
     }
 
-    /* create a dummy qp in order to get a unique qp_num to provide to librdmacm */
-    status = uct_rdmacm_cm_create_dummy_qp(cep->id, cq, &cep->qp);
+    if (supported) {
+        status = uct_rdamcm_cm_ep_create_reserved_qpn(cep);
+        if (status == UCS_OK) {
+            cep->flags |= UCT_RDMACM_CM_EP_QPN_CREATED;
+            *qpn = cep->qpn;
+        }        
+    } else {
+        status = uct_rdmacm_cm_get_cq(uct_rdmacm_cm_ep_get_cm(cep), cep->id->verbs, &cq);
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        /* create a dummy qp in order to get a unique qp_num to provide to librdmacm */
+        status = uct_rdmacm_cm_create_dummy_qp(cep->id, cq, &cep->qp);
+        if (status == UCS_OK) {
+            cep->flags |= UCT_RDMACM_CM_EP_QPN_CREATED;
+            *qpn = cep->qp->qp_num;
+        }
+    }
+
+    return status;
+}
+
+static void uct_rdamcm_cm_ep_destroy_qpn(uct_rdmacm_cm_ep_t *cep)
+{
+    ucs_status_t status;
+    bool supported;
+
+    if (!(cep->flags & UCT_RDMACM_CM_EP_QPN_CREATED)) {
+        return;
+    }
+
+    status = uct_rdamcm_cm_ep_is_reserved_qpn_supported(cep, &supported);
     if (status != UCS_OK) {
-        return status;
+        return;
     }
 
-    conn_param->qp_num = cep->qp->qp_num;
-    return UCS_OK;
+    if (supported) {
+        uct_rdamcm_cm_ep_destroy_reserved_qpn(cep);
+    } else {
+        uct_rdmacm_cm_ep_destroy_dummy_qp(cep);
+    }
```

Comment:
if implement above then it could be like this:
```
qpn_context_t *ctx = get_qpn_context(ep);
if (ctx->flag & RESERVED_QPN_SUPPORTED) {
    uct_rdamcm_cm_ep_destroy_reserved_qpn(cep);
} else {
    uct_rdmacm_cm_ep_destroy_dummy_qp(cep);
}
```

Reply:
@evgeny-leksikov Same reasons as above.

Reply:
@evgeny-leksikov Fixed.